compiler/rustc_hir/src/hir.rs

```diff
@@ -1931,6 +1931,7 @@ pub enum ExprKind<'hir> {
     /// This may also be a generator literal or an `async block` as indicated by the
     /// `Option<Movability>`.
     Closure {
+        binder: &'hir ClosureBinder<'hir>,
         capture_clause: CaptureBy,
         bound_generic_params: &'hir [GenericParam<'hir>],
```

Comment:
This field also holds generic parameters introduced by lifetime resolution.  Can this and `binder` be merged?

Reply:
This would require to distinguish somewhere the presence of the `for<>` in code vs just in HIR.

Reply:
Honestly, I'm not sure. This can probably be `binder: ClosureBinder { Present, NotPresent }` and all lifetimes stored in `bound_generic_params`, but this makes it impossible to distinguish lifetimes from binder from other lifetimes... No idea if we really need to distinguish them though

Reply:
`bound_generic_params` is only filled by anonymous lifetimes, that you explicitly ban.  So I don't think you need to worry about separating both.
Furthermore, if we ever allow anonymous lifetimes with a `for<>`, we will want them to use the `for<>` semantics and not the default/legacy one, won't we?

Reply:
So I've tried this in 27905df77fd6dd9ad6d73b7820ff3d020ec2e7b3 and... in my humble opinion the code turned out to be worse.